INT-4300: Add WebFlux server support

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4300

* Add `ReactiveHttpInboundEndpoint` based on the WebFlux foundation
* Extract `BaseHttpInboundEndpoint` for common
HTTP Inbound Channel Adapters options
* Make `spring-webmvc` and `spring-webflux` as `optional` dependencies
to let end-user to choose
* Refactor `HttpContextUtils` to include constants
for newly added WebFlux support
* Introduce `BaseHttpInboundEndpoint.setRequestPayloadTypeClass()`
for raw `Class<?>` and modify existing `setRequestPayloadType()`
for the `ResolvableType`
* Refactor existing MVC tests and XML components parsers to use
new `setRequestPayloadTypeClass()`
* Add `MessagingGatewaySupport.sendAndReceiveMessageReactive()` to get
a reply from downstream flow reactive back-pressure manner
* Add `IntegrationHandlerResultHandler` implementation to let WebFlux
infrastructure to handle the `Mono<Void>` from the `ReactiveHttpInboundEndpoint`
properly
* Fix `JdbcLockRegistryLeaderInitiatorTests` race condition to assert
the `initiator1` is elected eventually after yielding when the `initiator2`
is stopped

artembilan committed:
INT-4300: Add WebFlux server support

JIRA: https://jira.spring.io/browse/INT-4300

* Add `ReactiveHttpInboundEndpoint` based on the WebFlux foundation
* Extract `BaseHttpInboundEndpoint` for common
HTTP Inbound Channel Adapters options
* Make `spring-webmvc` and `spring-webflux` as `optional` dependencies
to let end-user to choose
* Refactor `HttpContextUtils` to include constants
for newly added WebFlux support
* Introduce `BaseHttpInboundEndpoint.setRequestPayloadTypeClass()`
for raw `Class<?>` and modify existing `setRequestPayloadType()`
for the `ResolvableType`
* Refactor existing MVC tests and XML components parsers to use
new `setRequestPayloadTypeClass()`
* Add `MessagingGatewaySupport.sendAndReceiveMessageReactive()` to get
a reply from downstream flow reactive back-pressure manner
* Add `IntegrationHandlerResultHandler` implementation to let WebFlux
infrastructure to handle the `Mono<Void>` from the `ReactiveHttpInboundEndpoint`
properly
* Fix `JdbcLockRegistryLeaderInitiatorTests` race condition to assert
the `initiator1` is elected eventually after yielding when the `initiator2`
is stopped

artembilan commented:
Really need tests and afterwards Docs.
But ready for feedback

Thanks

/CC @smaldini , @rstoyanchev

rstoyanchev commented on spring-integration-http/src/main/java/org/springframework/integration/http/inbound/ReactiveHttpInboundEndpoint.java:
```diff
@@ -0,0 +1,382 @@
+		return Mono.just(exchange)
+				.doOnSubscribe(s -> this.activeCount.incrementAndGet())
+				.flatMap(this::extractRequestBody)
+				.switchIfEmpty(Mono.just(exchange.getRequest().getQueryParams()))
```
Why `Mono.just` + `flatMap` vs just `extractRequestBody()`?

rstoyanchev replied:
Also the `switchIfEmpty` is probably better handled as part of `extractRequestBody` where there are some checks already for what kind of request it is. For example an HTTP POST with form data should always have a body and hence a fallback on query params doesn't make much sense. And do you really want query params as the body any time there is no body? I would think that only makes sense with something like a search form, i.e. HTTP GET with query params.

artembilan replied:
Thank you, @rstoyanchev , for feedback.
Yep! 
```
return extractRequestBody(exchange)
				.doOnSubscribe(s -> this.activeCount.incrementAndGet())
				.map(body -> buildMessage(new HttpEntity<>(body, exchange.getRequest().getHeaders()), exchange))
```
does exactly the same and looks better.
I have moved `switchIfEmpty()` logic to the `extractRequestBody()`.
We just use that query params fallback in the standard, blocking implementation:
```
if (payload == null) {
	if (requestBody != null) {
		payload = requestBody;
	}
	else {
		payload = requestParams;
	}
}
```

rstoyanchev commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -533,6 +543,127 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+							this.messageCount.incrementAndGet();
+						}
+					})
+					.flatMap(message -> {
```
Same here. The `Mono.just` + `flatMap` seems unnecessary. @smaldini ?

artembilan replied:
OK. I'm changing this combination to the `Mono.defer(() -> {`.
Thank you!

artembilan commented:
Ready for merge.

I don't see reason to add XML configuration support.
I'll add Docs eventually after some review feedback.

Also, I guess, we need Migration Guide note that MVC is already `optional`, because you may select WebFlux instead.

artembilan committed:
* Fix JavaDocs issue in the `HttpRequestHandlingMessagingGateway`
* Move all the "hard" logic in the `MessagingGatewaySupport#doSendAndReceiveMessageReactive`
to the `Mono` chain ensuring back-pressure when `sendAndReceiveMessageReactive()` is
called not from the Reactive Stream

artembilan committed:
Add test-case to demonstrate SSE

JIRA: https://jira.spring.io/browse/INT-3625

artembilan committed:
Some polishing and optimization for the
`MessagingGatewaySupport.doSendAndReceiveMessageReactive()`

smaldini commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -560,42 +700,56 @@ protected AttributeAccessor getErrorMessageAttributes(Message<?> message) {
+	protected void registerReplyMessageCorrelatorIfNecessary() {
+		MessageChannel replyChannel = getReplyChannel();
+		if (replyChannel != null && this.replyMessageCorrelator == null) {
+			synchronized (this.replyMessageCorrelatorMonitor) {
```
Review synchronized use:
when some code is synchronized, make sure is only doing local state mutation, then do call  user provided code outside, e.g. 
```boolean shouldStart = false;
    synchronized { //.. 
          shouldStart = isRunning();
     }
     if(shouldStart) {
        correlator.start();
     }
```

smaldini commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -533,6 +546,135 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+	private Mono<Message<?>> doSendAndReceiveMessageReactive(MessageChannel requestChannel, Object object,
+			boolean error) {
+
+		AtomicReference<Message<?>> requestMessageReference = new AtomicReference<>();
```
No need for this.
Move lines 623 `doOnSubscribe` and 628 `onErrorResume` inside the `defer`, just after line 621 `map`

artembilan replied:
Ah! Cool! 
Thanks 😄

smaldini commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -533,6 +546,135 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+							.setErrorChannel(replyChannel)
+							.build();
+
+					requestMessageReference.set(requestMessage);
```
No need if comment line 566 is implemented

smaldini commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -533,6 +546,135 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+						this.messageCount.incrementAndGet();
+					}
+				})
+				.onErrorResume(t -> handleSendError(requestMessageReference.get(), t));
```
if moved inside defer, can directly use message created in it.

smaldini commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java:
```diff
@@ -533,6 +546,135 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
+		if (errorChannel != null) {
+			ErrorMessage errorMessage = buildErrorMessage(requestMessage, exception);
+			try {
+				return doSendAndReceiveMessageReactive(errorChannel, errorMessage, true);
```
Does it resume the response from error channel then ? If so the message will be different, is it expected ?

artembilan replied:
That's correct. If error handling flow returns good result we sent it back to the subscriber.
If that fails with an exception, we just propagate it into `onError`.

This behavior is fully similar to what we have with standard, blocking `sendAndReceive()`

artembilan replied:
Thank you for the pointer, @smaldini !

After farther review I noticed that we really don't stop to call `handleSendError()` recursively.
That is exactly for what I have that `boolean error` argument.
So, now it is like this:
```
.onErrorResume(t -> error ? Mono.error(t) : handleSendError(requestMessage, t));
```

smaldini commented:
LGTM besides minor tweaks I commented. I'd need to diagnose an end to end flow when time permits to see where further support should be made separately 👍

artembilan committed:
More optimization for `MessagingGatewaySupport`

garyrussell commented:
>I'll add Docs eventually after some review feedback.

Docs?

artembilan commented:
> Docs?

Will do today.
Thanks

artembilan committed:
* Upgrade to Reactor 3.1 M3
* Document WebFlux-based components

garyrussell commented:
Minor doc polish here: https://github.com/garyrussell/spring-integration/commit/90cad0dcafbf7d8581237c7bbf9dfb46cad0f53b

garyrussell commented on build.gradle:
```diff
@@ -391,8 +391,9 @@ project('spring-integration-http') {
-		compile "org.springframework:spring-webflux:$springVersion"
+		compile "org.springframework:spring-web:$springVersion"
+		compile ("org.springframework:spring-webmvc:$springVersion", optional)
+		compile ("org.springframework:spring-webflux:$springVersion", optional)
```
Perhaps we should split into 3 modules; `http-common`, `http-mvc` (or leave it as `http`) and `http-webflux`, to make the transitive dependencies work ok for users.

garyrussell replied:
It can be in the next milestone, however.

artembilan replied:
Well, actually I'm starting to think that we can make them all non-optional.
As I see by our tests it doesn't hurt to have MVC and WebFlux in the same CP.
That is in the end only end-user choice which engine to choose: `@EnableWebMvc` or `@EnableWebFlux` and configure particular server  - Servlet Container or Reactor.

garyrussell replied:
I don't think we want add webflux and reactor onto the CP if they're only using MVC.

artembilan replied:
OK. So be it as `spring-integration-webflux` then!
We can make it depend on SI-http and exclude MVC.

Next milestone ?

garyrussell replied:
OK; please open a JIRA so we don't forget. 

I don't think exclude is right though; it is possible they want both.

artembilan replied:
https://jira.spring.io/browse/INT-4315

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/d4a99919ed78c011dc8d69bae745eedd425eb672
